precompiles/addr/addr_test.go

```diff
@@ -0,0 +1,215 @@
+	k.SetAddressMapping(ctx, callerSeiAddress, callerEvmAddress)
+	callerSig, err := crypto.Sign(hash.Bytes(), callerKey)
+	callerR := fmt.Sprintf("0x%v", new(big.Int).SetBytes(callerSig[:32]).Text(16))
+	callerS := fmt.Sprintf("0x%v", new(big.Int).SetBytes(callerSig[32:64]).Text(16))
+	callerV := fmt.Sprintf("0x%v", new(big.Int).SetBytes([]byte{callerSig[64]}).Text(16))
+
+	happyPathOutput, _ := associateWithGas.Outputs.Pack(addr.AddrPair{SeiAddr: targetSeiAddress.String(), EvmAddr: targetEvmAddress})
+
+	type args struct {
+		evm    *vm.EVM
+		caller common.Address
+		v      string
+		r      string
+		s      string
+		msg    string
+		value  *big.Int
+	}
+	tests := []struct {
+		name       string
+		args       args
+		wantRet    []byte
+		wantErr    bool
+		wantErrMsg string
+		wrongRet   bool
+	}{
+		{
+			name: "fails if payable",
+			args: args{
+				evm: &vm.EVM{
+					StateDB:   state.NewDBImpl(ctx, k, true),
+					TxContext: vm.TxContext{Origin: callerEvmAddress},
+				},
+				caller: callerEvmAddress,
+				v:      v,
+				r:      r,
+				s:      s,
+				msg:    prefixedMessage,
+				value:  big.NewInt(10),
+			},
+			wantRet:    nil,
+			wantErr:    true,
+			wantErrMsg: "sending funds to a non-payable function",
+			wrongRet:   false,
+		},
+		{
+			name: "fails if input is not hex",
+			args: args{
+				evm: &vm.EVM{
+					StateDB:   state.NewDBImpl(ctx, k, true),
+					TxContext: vm.TxContext{Origin: callerEvmAddress},
+				},
+				caller: callerEvmAddress,
+				v:      "nothex",
+				r:      r,
+				s:      s,
+				msg:    prefixedMessage,
+				value:  big.NewInt(0),
+			},
+			wantRet:    nil,
+			wantErr:    true,
+			wantErrMsg: "encoding/hex: invalid byte: U+006E 'n'",
+			wrongRet:   false,
+		},
+		{
+			name: "fails if addresses are already associated",
+			args: args{
+				evm: &vm.EVM{
+					StateDB:   state.NewDBImpl(ctx, k, true),
+					TxContext: vm.TxContext{Origin: callerEvmAddress},
+				},
+				caller: callerEvmAddress,
+				v:      callerV,
+				r:      callerR,
+				s:      callerS,
+				msg:    prefixedMessage,
+				value:  big.NewInt(0),
+			},
+			wantRet:    nil,
+			wantErr:    true,
+			wantErrMsg: fmt.Sprintf("address %s is already associated with evm address %s", callerSeiAddress, callerEvmAddress),
+		},
+		{
+			name: "associates wrong address if invalid signature (different values)",
+			args: args{
+				evm: &vm.EVM{
+					StateDB:   state.NewDBImpl(ctx, k, true),
+					TxContext: vm.TxContext{Origin: callerEvmAddress},
+				},
+				caller: callerEvmAddress,
+				v:      v,
+				r:      r,
+				s:      r, // Pass in r instead of s here for invalid value
+				msg:    prefixedMessage,
+				value:  big.NewInt(0),
+			},
+			wantRet:    happyPathOutput,
+			wantErr:    false,
+			wantErrMsg: "",
+			wrongRet:   true,
+		},
+		{
+			name: "associates wrong address if invalid signature (different message)",
```

Comment:
Curious about this test case. Do we expect association if signature is invalid?

Reply:
Yup in this case we try to derive a pubkey from the signature values and signed message (which seems to be always possible), it'll just give us the wrong pubkey.